core/Tracker/Failures.php

```diff
@@ -0,0 +1,178 @@
+        // eg imagine a super user tries to issue an API request for a site and sending the wrong parameter for a token...
+        // an admin may have view access for this and can see the super users token
+        $token = $request->getTokenAuth();
+        $params = $request->getRawParams();
+        foreach (array('token_auth', 'token', 'tokenauth', 'token__auth') as $key) {
+            if (isset($params[$key])) {
+                $params[$key] = '__ANONYMIZED__';
+            }
+        }
+        foreach ($params as $key => $value) {
+            if (!empty($token) && $value === $token) {
+                $params[$key] = '__ANONYMIZED__'; // user accidentally posted the token in a wrong field
+            } elseif (!empty($value) && is_string($value) && Common::mb_strlen($value) >= 29 && Common::mb_strlen($value) <= 36 && ctype_xdigit($value)) {
+                $params[$key] = '__ANONYMIZED__'; // user maybe posted a token in a different field... it looks like it might be a token
+            }
+        }
+
+        return $params;
+    }
+
+    public function removeFailuresOlderThanDays($days)
+    {
+        $minutesAgo = $this->getNow()->subDay($days)->getDatetime();
+
+        PiwikDb::query(sprintf('DELETE FROM %s WHERE date_first_occurred < ?', $this->tablePrefixed), array($minutesAgo));
+    }
+
+    public function getAllFailures()
+    {
+        $failures = PiwikDb::fetchAll(sprintf('SELECT * FROM %s', $this->tablePrefixed));
+        return $this->enrichFailures($failures);
+    }
+
+    public function getFailuresForSites($idSites)
+    {
+        if (empty($idSites)) {
+            return array();
+        }
+        $idSites = array_map('intval', $idSites);
+        $idSites = implode(',', $idSites);
+        $failures = PiwikDb::fetchAll(sprintf('SELECT * FROM %s WHERE idsite IN (%s)', $this->tablePrefixed, $idSites));
+        return $this->enrichFailures($failures);
+    }
+
+    public function deleteTrackingFailure($idSite, $idFailure)
+    {
+        PiwikDb::query(sprintf('DELETE FROM %s WHERE idsite = ? and idfailure = ?', $this->tablePrefixed), array($idSite, $idFailure));
+    }
+
+    public function deleteTrackingFailures($idSites)
+    {
+        if (!empty($idSites)) {
+            $idSites = array_map('intval', $idSites);
+            $idSites = implode(',', $idSites);
+            PiwikDb::query(sprintf('DELETE FROM %s WHERE idsite IN(%s)', $this->tablePrefixed, $idSites));
+        }
+    }
+
+    public function deleteAllTrackingFailures()
+    {
+        PiwikDb::query(sprintf('DELETE FROM %s', $this->tablePrefixed));
+    }
+
+    private function enrichFailures($failures)
+    {
+        foreach ($failures as &$failure) {
+            try {
+                $failure['site_name'] = Site::getNameFor($failure['idsite']);
+            } catch (UnexpectedWebsiteFoundException $e) {
+                $failure['site_name'] = Piwik::translate('General_Unknown');
+            }
```

Comment:
If a site is not known would it be more useful to have the idSite displayed in the name? Not sure if the idSite is also displayed in the UI.

Reply:
Nevermind, see the screenshot below.